### PR TITLE
Feature AB#12 AB#11 :[Add Auto mapper , Mediatr,StyleCop ]

### PR DESCRIPTION
**Pull request by MarwenSaad**
Feature [AB#12](https://dev.azure.com/TunNetCom/3348a3bb-788f-42b4-a0ab-ba7210c86b88/_workitems/edit/12) [AB#11](https://dev.azure.com/TunNetCom/3348a3bb-788f-42b4-a0ab-ba7210c86b88/_workitems/edit/11) :[Add Auto mapper , Mediatr,StyleCop ]

### CODE REVIEW

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+{
+    using AutoMapper;
+    using TunNetCom.AionTime.TimeLogService.Contracts.DTOs.Request;
+    using TunNetCom.AionTime.TimeLogService.Domain.Models;
+
+    public class MappingProfile: Profile
+    {
+        public MappingProfile()
+        {
+            this.CreateMap<OrganizationRequest, Organization>().ReverseMap();
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:11]: no need for the ReverseMap

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Application.Feature.OrganizationAction.Commands.AddOrganization;
+
+public class AddOrganizationCommandHandler(IGenericRepository<Organization> organizationRepository, IMapper mapper) : IRequestHandler<AddOrganizationCommand, int>
+{
+    private readonly IGenericRepository<Organization> organizationRepository = organizationRepository;
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommandHandler.cs:5]: IGenericRepository  --> IRepository

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Application.AutoMapping
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:1]: ﻿namespace TunNetCom.AionTime.TimeLogService.Application.AutoMapping  --->  ﻿namespace TunNetCom.AionTime.TimeLogService.Application.AutoMapping;

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs [hunk 4]
```diff
@@ -0,0 +1,2 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Application.Feature.OrganizationAction.Commands;
+public record AddOrganizationCommand(OrganizationRequest organizationDTO) : IRequest<int>;
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs:2]: we can't accept a DTO in the command

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    private readonly IMapper mapper = mapper;
+
+    async Task<int> IRequestHandler<AddOrganizationCommand, int>.Handle(AddOrganizationCommand request, CancellationToken cancellationToken)
+    {
+        var organization = this.mapper.Map<Organization>(request.organizationDTO);
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommandHandler.cs:10]: we need to implement the DDD

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Application.Feature.OrganizationAction.Commands.DeletOrganization;
+
+public class DeleteOrganizationCommandHandler(IGenericRepository<Organization> organizationRepository, IMapper mapper) : IRequestHandler<DeleteOrganizationCommand, int>
+{
+    private readonly IMapper mapper=mapper;
+    private readonly IGenericRepository<Organization> organizationRepository = organizationRepository;
+
+    public async Task<int> Handle(DeleteOrganizationCommand request, CancellationToken cancellationToken)
+    {
+        var organization = this.mapper.Map<Organization>(request.organization);
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs:10]: in the delete we need just the Id

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Queries/GetOrganization/GetOrganizationsQueryHandler.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Queries/GetOrganization/GetOrganizationsQueryHandler.cs:1]: unused usings

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Queries/GetOrganization/GetOrganizationsQueryHandler.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class GetOrganizationsQueryHandler(IGenericRepository<Organization> organizationRepository, IMapper mapper) : IRequestHandler<GetOrganizationsQuery, IReadOnlyList<OrganizationRequest>>
+    {
+        private readonly IGenericRepository<Organization> organizationRepository = organizationRepository;
+        private readonly IMapper mapper = mapper;
+
+        public async Task<IReadOnlyList<OrganizationRequest>> Handle(GetOrganizationsQuery request, CancellationToken cancellationToken)
+        {
+            var res = await this.organizationRepository.GetAsync();
+            var organizations =this.mapper.Map<IReadOnlyList<OrganizationRequest>>(res);
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Queries/GetOrganization/GetOrganizationsQueryHandler.cs:17]: dont forget the spaces

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    private readonly IMediator mediator;
+
+    public TimeLogServiceController(
+        ILogger<WorkItemTimeLog> logger,
+        IMediator mediator)
+    {
+        this.logger = logger;
+        this.mediator = mediator;
+    }
+
```
Comment (Nieze-BenMansour) [inline src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs:17]: Ya A5I mayjish el controller kolhom yraj3ou ok, check this: 
https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design

**Review by Nieze-BenMansour — CHANGES_REQUESTED** [review]

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs (hunk 1 above)
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:11]: I need this for the GetOrganizationHandler to reverse mapping

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs (hunk 4 above)
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs:2]: Entity ?

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs (hunk 6 above)
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs:10]: public async Task DeleteAsync(T entity)
{
    _ = this.context.Remove(entity);
    _ = await this.context.SaveChangesAsync();
}

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs (hunk 9 above)
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs:17]: I will implement a global exception handler

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs (hunk 1 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:11]: use "File scoped namespaces"

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs (hunk 3 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:1]: use "File scoped namespaces"

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs (hunk 1 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/AutoMapping/MappingProfile.cs:11]: Entites should be made by their parametred constructors

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs (hunk 4 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/AddOrganization/AddOrganizationCommand.cs:2]: nope parameters

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs (hunk 6 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/DeletOrganization/DeleteOrganizationCommandHandler.cs:10]: bad implementation and remebrer about the soft delete

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs (hunk 9 above)
Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs:17]: working with exception will cost us a lot of memory leaks

Comment (Nieze-BenMansour) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Controllers/TimeLogServiceController.cs:17]: and we should follow the rest standart

**Review by Nieze-BenMansour — CHANGES_REQUESTED** [review]

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Contracts/CustomProblemDetails/CustomProblemDetails.cs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Contracts;
+using Microsoft.AspNetCore.Mvc;
+public class CustomProblemDetails : ProblemDetails
+{
+    public IDictionary<string, string> Errors { get; set; } = new Dictionary<string, string>();
```
[the remark this answers is not in the document]
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Contracts/CustomProblemDetails/CustomProblemDetails.cs:5]: the natve one contain only 

 Title = 
 Status =
 Detail = 
 Type =

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Middelware/GlobalExceptionHandler.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+                    Type = exception.GetType().Name,
+                }; 
+                break;
+
+        }
+
+
+        httpContext.Response.StatusCode = (int)statusCode;
+        string logMessage = JsonConvert.SerializeObject(problem);
+        this.logger.LogError(logMessage);
```
[the remark this answers is not in the document]
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Budget.API/Middelware/GlobalExceptionHandler.cs:53]: ????

File: src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/UpdateOrganization/UpdateOrganizationCommandHandler.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿namespace TunNetCom.AionTime.TimeLogService.Application;
+
+public class UpdateOrganizationCommandHandler(IRepository<Organization> organizationRepository, IMapper mapper) : IRequestHandler<UpdateOrganizationCommand, int>
+{
+    private readonly IRepository<Organization> organizationRepository = organizationRepository;
+    private readonly IMapper mapper = mapper;
+
+    public async Task<int> Handle(UpdateOrganizationCommand request, CancellationToken cancellationToken)
+    {
+        var organization = this.mapper.Map<Organization>(request.organization);
```
[the remark this answers is not in the document]
Comment (MarwenSaad) [reply src/TunNetCom.AionTime.TimeLogService/TunNetCom.AionTime.TimeLogService.Application/Feature/OrganizationAction/Commands/UpdateOrganization/UpdateOrganizationCommandHandler.cs:10]: ??

**Pull request merged**